test_conformance/basic/test_intmath.cpp

```diff
@@ -0,0 +1,240 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <functional>
+#include <string>
+#include <vector>
+
+#include "procs.h"
+
+template <typename T> struct TestDef
+{
+    const char *name;
+    const char *kernel_code;
+    std::function<T(T, T, T)> ref;
+};
+
+template <typename T, unsigned N>
+int test_intmath(cl_device_id device, cl_context context,
+                 cl_command_queue queue, int num_elements, std::string typestr)
+{
+    TestDef<T> tests[] = {
+        // Test addition
+        {
+            "test_add",
+            R"(
+  __kernel void test_add(__global TYPE *srcA,
+                         __global TYPE *srcB,
+                         __global TYPE *srcC,
+                         __global TYPE *dst)
+  {
+      int  tid = get_global_id(0);
+      dst[tid] = srcA[tid] + srcB[tid];
+  };
+)",
+            [](T a, T b, T c) { return a + b; },
+        },
+
+        // Test subtraction
+        {
+            "test_sub",
+            R"(
+  __kernel void test_sub(__global TYPE *srcA,
+                         __global TYPE *srcB,
+                         __global TYPE *srcC,
+                         __global TYPE *dst)
+  {
+      int  tid = get_global_id(0);
+      dst[tid] = srcA[tid] - srcB[tid];
+  };
+)",
+            [](T a, T b, T c) { return a - b; },
+        },
+
+        // Test multiplication
+        {
+            "test_mul",
+            R"(
+  __kernel void test_mul(__global TYPE *srcA,
+                         __global TYPE *srcB,
+                         __global TYPE *srcC,
+                         __global TYPE *dst)
+  {
+      int  tid = get_global_id(0);
+      dst[tid] = srcA[tid] * srcB[tid];
+  };
+)",
+            [](T a, T b, T c) { return a * b; },
+        },
+
+        // Test multiply-accumulate
+        {
+            "test_mad",
+            R"(
+  __kernel void test_mad(__global TYPE *srcA,
+                         __global TYPE *srcB,
+                         __global TYPE *srcC,
+                         __global TYPE *dst)
+  {
+      int  tid = get_global_id(0);
+      dst[tid] = srcA[tid] * srcB[tid] + srcC[tid];
+  };
+)",
+            [](T a, T b, T c) { return a * b + c; },
+        },
+    };
+
+    clMemWrapper streams[4];
+    cl_int err;
+
+    if (std::is_same<T, cl_ulong>::value && !gHasLong)
+    {
+        log_info("64-bit integers are not supported on this device. Skipping "
+                 "test.\n");
+        return TEST_SKIP;
```

Comment:
This is quirky and we should fix the harness but the correct way of doing this at the moment is to return `TEST_SKIPPED_ITSELF`.